<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [97]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 


In [98]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [106]:
connection.close()

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [13]:
# текст запроса
query_3_1 = f'''select COUNT(*)
                from public.vacancies
            '''

In [14]:
# результат запроса
query_3_1_result = pd.read_sql_query(query_3_1, connection)
query_3_1_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [15]:
# текст запроса
query_3_2 = f'''select COUNT(*)
                from public.employers
            '''

In [16]:
# результат запроса
query_3_2_result = pd.read_sql_query(query_3_2, connection)
query_3_2_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [17]:
# текст запроса
query_3_3 = f'''select COUNT(*)
                from public.areas
            '''

In [18]:
# результат запроса
query_3_3_result = pd.read_sql_query(query_3_3, connection)
query_3_3_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [19]:
# текст запроса
query_3_4 = f'''select COUNT(*)
                from public.industries
            '''

In [20]:
# результат запроса
query_3_4_result = pd.read_sql_query(query_3_4, connection)
query_3_4_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

***

In [ ]:
# выводы по предварительному анализу данных

- Number of vacancies = 49 197
- Number of employers = 23 501
- Number of areas = 1 362
- Number of industries = 294

- These numbers make sense and seem to reflect the real world scenario. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [21]:
# текст запроса
query_4_1 = f'''select
                    a.name area,
                    COUNT(*) cnt
                from public.vacancies v
                    join public.areas a on v.area_id = a.id
                group by a.id
                order by 2 desc
            '''

In [22]:
# результат запроса
query_4_1_result = pd.read_sql_query(query_4_1, connection)
query_4_1_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area   cnt
0                      Москва  5333
1             Санкт-Петербург  2851
2                       Минск  2112
3                 Новосибирск  2006
4                      Алматы  1892
..                        ...   ...
764                    Кизляр     1
765                    Джизак     1
766                    Эртиль     1
767                  Арсеньев     1
768  Тура (Красноярский край)     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [23]:
# текст запроса
query_4_2 = f'''select 
                    COUNT(*)
                from public.vacancies
                where salary_from IS NOT NULL
                    OR salary_to IS NOT NULL
            '''

In [24]:
# результат запроса
query_4_2_result = pd.read_sql_query(query_4_2, connection)
query_4_2_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [27]:
# текст запроса 1
query_4_3 = f'''select 
                    AVG(salary_from)
                from public.vacancies
            '''

In [28]:
# результат запроса 1
query_4_3_result = pd.read_sql_query(query_4_3, connection)
round(query_4_3_result, 0)

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg
0  71065.0

In [29]:
# текст запроса 2
query_4_3 = f'''select 
                    AVG(salary_to)
                from public.vacancies
            '''

In [30]:
# результат запроса 2
query_4_3_result = pd.read_sql_query(query_4_3, connection)
round(query_4_3_result, 0)

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg
0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [31]:
# текст запроса
query_4_4 = f'''select 
                    schedule,
                    employment,
                    count(*)
                from public.vacancies
                group by 1, 2
                order by 3 desc
            '''

In [32]:
# результат запроса
query_4_4_result = pd.read_sql_query(query_4_4, connection)
query_4_4_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [33]:
# текст запроса
query_4_5 = f'''select 
                    experience,
                    COUNT(*)
                from public.vacancies
                group by 1
                order by 2
            '''

In [34]:
# результат запроса
query_4_5_result = pd.read_sql_query(query_4_5, connection)
query_4_5_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [ ]:
# выводы по детальному анализу вакансий

- Мы наблюдаем самое большое количество вакансий в Москве (5 333).

- Почти у половины вакансий не заполнено одно из двух полей с зарплатой, что показывает на весомое количество пропусков. Возможно, это говорит о том, что рынок труда сверхдинамичен, соответственно, точные границы оплаты труда не могут быть выставлены.

- Нижняя граница среднего значения зарплатной вилки = 71 000 руб
- Верхняя граница среднего значения зарплатной вилки  = 110 000 руб
- Предположительно, это репрезентативно реальной ситуации на рынке труда.

- Полный рабочий день + полная занятость является самым востребованным сочетанием типов трудоустройства (35 367 вакансий).

- Категория опыта работы от 1 года до 3 лет имет наибольшее количество вакансий (26152).

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [35]:
# текст запроса
query_5_1 = f'''select 
                    e.name,
                    COUNT(*)
                from public.employers e
                     join public.vacancies v on e.id = v.employer_id
                group by e.id
                order by 2 desc
            '''

In [36]:
# результат запроса
query_5_1_result = pd.read_sql_query(query_5_1, connection)
query_5_1_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0                           Яндекс   1933
1                       Ростелеком    491
2                         Тинькофф    444
3                             СБЕР    428
4                    Газпром нефть    331
...                            ...    ...
14901   Евсеев Сергей Владимирович      1
14902          Кадровые технологии      1
14903                   ЛОМБАРД 24      1
14904            Авто-Компонент ТД      1
14905  НПП «Старт» им. А.И. Яскина      1

[14906 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [40]:
# текст запроса
query_5_2 = f'''select 
                    a.name area,
                    count(distinct e.id) employers_cnt,
                    count(distinct v.id) vacancies_cnt
                from 
                    public.areas a
                    left join public.vacancies v on a.id = v.area_id
                    left join public.employers e on a.id = e.area
                where v.id is null
                group by a.id
                order by 2 desc nulls first
                '''

In [41]:
# результат запроса
query_5_2_result = pd.read_sql_query(query_5_2, connection)
query_5_2_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area  employers_cnt  vacancies_cnt
0                Россия            410              0
1             Казахстан            207              0
2    Московская область             75              0
3    Краснодарский край             19              0
4    Ростовская область             18              0
..                  ...            ...            ...
588             Скидель              0              0
589          Белоозерск              0              0
590           Орловский              0              0
591          Павловская              0              0
592             Таксимо              0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [38]:
# текст запроса
query_5_3 = f'''select 
                    employer_id,
                    count(distinct area_id)
                from public.vacancies
                group by 1
                order by 2 desc
            '''

In [39]:
# результат запроса
query_5_3_result = pd.read_sql_query(query_5_3, connection)
query_5_3_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer_id  count
0             1740    181
1             2748    152
2          5724811    116
3          5130287     88
4          3682876     71
...            ...    ...
14901       810278      1
14902       810313      1
14903       810551      1
14904       810688      1
14905       811902      1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [42]:
# текст запроса
query_5_4 = f'''select 
                    count(*)
                from 
                    public.employers e
                    left join public.employers_industries i on e.id = i.employer_id
                where i.industry_id is null
            '''

In [43]:
# результат запроса
query_5_4_result = pd.read_sql_query(query_5_4, connection)
query_5_4_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [44]:
# текст запроса
query_5_5 = f'''select
                    e.name,
                    count(i.industry_id)
                from 
                    public.employers e
                    left join public.employers_industries i on e.id = i.employer_id
                group by e.id
                having count(i.industry_id) = 4
                order by 1
                offset 2 limit 1
            '''

In [45]:
# результат запроса
query_5_5_result = pd.read_sql_query(query_5_5, connection)
query_5_5_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [50]:
# текст запроса
query_5_6 = f'''select count(*)
                from 
                    public.employers e
                    join public.employers_industries i on e.id = i.employer_id
                    join public.industries ii on i.industry_id = ii.id
                where ii.name = 'Разработка программного обеспечения'
            '''

In [51]:
# результат запроса
query_5_6_result = pd.read_sql_query(query_5_6, connection)
query_5_6_result

count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [81]:
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url)
html = response.content
soup = BeautifulSoup(html, 'html.parser')
table = soup.find('table', {'class': 'standard sortable'})
rows = table.find_all('tr')

cities = []
for row in rows[1:]: 
    cells = row.find_all('td')
    if len(cells) > 0:
        city_name = cells[1].text.strip()
        cities.append(city_name)

cities = tuple(cities)
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [82]:
# текст запроса
query_5_7 = f'''select 
                    a.name,
                    count(*) cnt
                from 
                    public.areas a
                    join public.vacancies v on a.id = v.area_id
                    join public.employers e on v.employer_id = e.id
                where e.name = 'Яндекс' 
                    and a.name IN {cities}
                group by 1

                union all

                select 
                    'Total',
                    count(*)
                from 
                    public.areas a
                    join public.vacancies v on a.id = v.area_id
                    join public.employers e on v.employer_id = e.id
                where e.name = 'Яндекс' 
                    and a.name IN {cities}
                order by 2
            '''

In [83]:
# результат запроса
query_5_7_result = pd.read_sql_query(query_5_7, connection)
query_5_7_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

In [ ]:
# выводы по анализу работодателей

- Яндекс является лидером по числу вакансий - 1 933 вакансии по сравнению со вторым местом, которое занимает Ростелеком 491.

- Данные показывают, что в России и в Казахстане есть работодатели, но нет вакансий. Так как мы уже знаем, что в таких городах, как Москва и Алматы, есть вакансии, следовательно, мы можем предположить, что мы имеем ошибки в наших даных, а именно ошибки в числе работодателей, например в России и в Казахстане.

- У весьма большого количества работодателей (8 419) не указана сфера деятельности.

- Примерно 1/4 всех вакансий Яндекса (485) представлены в городах-миллиониках, с самым большим числом вакансий в Москве (54).

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [85]:
# текст запроса
query_6_1 = f'''select count(*)
                from public.vacancies
                where name ilike '%data%'
                    or name ilike '%данн%'
            '''

In [86]:
# результат запроса
query_6_1_result = pd.read_sql_query(query_6_1, connection)
query_6_1_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [87]:
# текст запроса
query_6_2 = f'''select count(*)
                from public.vacancies
                where (name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or (name not ilike '%HTML%' and name ilike '%ML%')
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%%')
                    AND
                    (name ilike '%junior%'
                    or experience ilike '%Нет опыта%'
                    or employment ilike '%Стажировка%')
            '''

In [88]:
# результат запроса
query_6_2_result = pd.read_sql_query(query_6_2, connection)
query_6_2_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [89]:
# текст запроса
query_6_3 = f'''select count(*)
                from public.vacancies
                where (name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or (name not ilike '%HTML%' and name ilike '%ML%')
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%%')
                    and
                    (key_skills ilike '%SQL%' 
                    or key_skills ilike '%postgres%')
            '''

In [90]:
# результат запроса
query_6_3_result = pd.read_sql_query(query_6_3, connection)
query_6_3_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [91]:
# текст запроса
query_6_4 = f'''select count(*)
                from public.vacancies
                where (name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or (name not ilike '%HTML%' and name ilike '%ML%')
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%%')
                    and
                    key_skills ilike '%python%'
            '''

In [92]:
# результат запроса
query_6_4_result = pd.read_sql_query(query_6_4, connection)
query_6_4_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [95]:
# текст запроса
query_6_5 = f'''select round(avg(length(key_skills) - length(replace(key_skills,CHR(9),''))+1), 2)
                from vacancies v
                where 
                    (lower(v.name) like '%data scientist%' 
                    or lower(v.name) like '%data science%' 
                    or lower(v.name) like '%исследователь данных%' 
                    or ((v.name) like '%ML%' and lower(v.name) not like '%html%') 
                    or lower(v.name) like '%machine learning%' 
                    or lower(v.name) like '%машинн%обучен%') 
                    and v.key_skills is not null
            '''

In [96]:
# результат запроса
query_6_5_result = pd.read_sql_query(query_6_5, connection)
query_6_5_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [99]:
# текст запроса
query_6_6 = f'''SELECT
                    experience,
                    ROUND(AVG(
                        (COALESCE(salary_from, 0) + COALESCE(salary_to, 0)) / 
                        CASE 
                            WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN 2
                            ELSE 1
                        END
                    )) AS average_salary
                FROM
                    public.vacancies
                WHERE
                    (name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or (name not ilike '%HTML%' and name like '%ML%')
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%%')
                    and (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                GROUP BY 1
            '''

In [100]:
# результат запроса
query_6_6_result = pd.read_sql_query(query_6_6, connection)
query_6_6_result

/Users/gleb.khamin/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  average_salary
0           Нет опыта         74643.0
1  От 1 года до 3 лет        139675.0
2       От 3 до 6 лет        243115.0

***

In [ ]:
# выводы по предметному анализу

- 1 771 вакансий имеют отношение к данным, что есть 3.6% от общего числа вакансий.

- Только около 3% от этих вакансий подходят для начинающего ДС специалиста, что есть критично мало.

- 229 вакансий указывают SQL и Postgres, как ключевые навыки.

- Python, как ключевой навык, указан только в 20% всех вакансий, что есть удивительно низко.

- В среднем 6 требуемых ключевых навыков указано в вакансиях для ДС специалиста.

- Средняя заработная плата специалистов ДС высока для, например, российского рынка труда. Специалист ДС без опыта работы может в среднем получать 74 643 руб, с опытом работы от 3 до 6 лет - 243 115 руб.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

- Суммарное число всех вакансий - 49 197. Так как мы имеем 23 501 работодателей, можно сказать, что в среднем на каждого работодателя приходится по две вакансии.

- Город-лидер по количеству вакансий - Москва (5 333 вакансии, что есть примерно 11% от общего числа вакансий).

- Яндекс является лидером по числу вакансий - 1 933 вакансии по сравнению со вторым местом, которое занимает Ростелеком 491.  Примерно 1/4 всех вакансий Яндекса (485) представлены в городах-миллиониках, с самым большим числом вакансий в Москве (54). Из этого мы можем сделать предположение, что рынок труда в городах-миллионниках более развит и благоприятен для поиска высокооплачиваемой работы, в том числе и работы в секторе IT.

- Наши данные говорят о том, что рынок труда сверхдинамичен, и точные границы оплаты труда не могут быть выставлены почти у половины вакансий.

- Судя по нашим данным, ДС специальность в таки странах, как Россия и Казахстан, находится на этапе развития - мы видим только 3.6% вакансий ДС от общего числа представленных вакансий. Также мы видим, что только 3% от этих вакансий подходят для начинающего ДС специалиста, что есть критично мало, и должно претерпеть изменения для успешного развития ДС индустрии в таких странах, как Россия и Казахстан.

- Оказалось, что Python, как ключевой навык, указан только в 20% всех вакансий, что лично для меня, есть удивительно мало.

- Не смотря на то, что индустрия ДС в странах, по которым мы имееем данные, только развивается, средняя заработная плата уже находится на весьма высоком уровне, например, российского рынка труда - специалист ДС без опыта работы может в среднем получать 74 643 руб, с опытом работы от 3 до 6 лет - 243 115 руб.